src/telemetry/opentelemetry.ts

```diff
@@ -0,0 +1,233 @@
+import { TraceExporter } from '@google-cloud/opentelemetry-cloud-trace-exporter';
+import { MetricExporter } from '@google-cloud/opentelemetry-cloud-monitoring-exporter';
+import { CloudPropagator } from '@google-cloud/opentelemetry-cloud-trace-propagator';
+
+import { containerDetector } from '@opentelemetry/resource-detector-container';
+import { gcpDetector } from '@opentelemetry/resource-detector-gcp';
+
+import dc from 'node:diagnostics_channel';
+const channel = dc.channel('fastify.initialization');
+
+const isProd = process.env.NODE_ENV === 'production';
+const isTest = process.env.NODE_ENV === 'test';
+const resourceDetectors = [
+  resources.envDetectorSync,
+  resources.hostDetectorSync,
+  resources.osDetectorSync,
+  resources.processDetectorSync,
+  containerDetector,
+  gcpDetector,
+];
+
+const traceExporter = isProd
+  ? new TraceExporter()
+  : new OTLPTraceExporter({
+      // hostname: 'jaeger-collector',
+      url: `http://${process.env.OTLP_COLLECTOR_HOST}:${process.env.OTLP_COLLECTOR_PORT}`,
+    });
+
+const spanProcessor = isProd
+  ? new node.BatchSpanProcessor(traceExporter)
+  : new node.SimpleSpanProcessor(traceExporter);
+
+const metricReader = isProd
+  ? new metrics.PeriodicExportingMetricReader({
+      exportIntervalMillis: 10_000,
+      exporter: new MetricExporter(),
+    })
+  : new PrometheusExporter({}, () => {
+      const { endpoint, port } = PrometheusExporter.DEFAULT_OPTIONS;
+      console.log(`metrics endpoint: http://localhost:${port}${endpoint}`);
+    });
+
+// Try to get the app version from the header, then query param, then default to unknown
+export const getAppVersion = (req: FastifyRequest): string => {
+  return req.headers['x-app-version'] || req.query['v'] || 'unknown';
+};
+
+export const SemanticAttributes = {
+  ...SemAttr,
+  DAILY_APPS_VERSION: 'dailydev.apps.version',
+  DAILY_APPS_USER_ID: 'dailydev.apps.userId',
+  DAILY_HEIMDALL_SESSION: 'dailydev.heimdall.session',
+};
+
+export const addApiSpanLabels = (
+  span: api.Span,
+  req: FastifyRequest,
+  // TODO: see if we want to add some attributes from the response
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  res?: FastifyReply,
+): void => {
+  span.setAttributes({
+    [SemanticAttributes.DAILY_APPS_VERSION]: getAppVersion(req),
+    [SemanticAttributes.DAILY_APPS_USER_ID]: req.userId || 'unknown',
+    [SemanticAttributes.DAILY_HEIMDALL_SESSION]:
+      req.cookies['das'] || 'unknown',
+  });
+};
+
+export const addPubsubSpanLabels = (
+  span: api.Span,
+  subscription: string,
+  message: Message | { id: string; data?: Buffer },
+): void => {
+  span.setAttributes({
+    [SemanticAttributes.MESSAGING_SYSTEM]: 'pubsub',
+    [SemanticAttributes.MESSAGING_DESTINATION]: subscription,
+    [SemanticAttributes.MESSAGING_MESSAGE_ID]: message.id,
+    [SemanticAttributes.MESSAGING_MESSAGE_PAYLOAD_SIZE_BYTES]:
+      message.data.length,
```

Comment:
In my experience, the uniqueness of these attributes will make it super expensive. I recommend removing it

Reply:
Sorry I was thinking of metrics, not tracing. You can ignore me